command/meta_providers.go

```diff
@@ -152,8 +152,35 @@ func (m *Meta) providerInstallSource() getproviders.Source {
 	return m.ProviderSource
 }
 
-// providerDevOverrideWarnings returns a diagnostics that contains at least
-// one warning if and only if there is at least one provider development
+// providerDevOverrideInitWarnings returns a diagnostics that contains at
+// least one warning if and only if there is at least one provider development
+// override in effect. If not, the result is always empty. The result never
+// contains error diagnostics.
+//
+// The init command can use this to include a warning that the results
+// may differ from what's expected due to the development overrides. For
+// other commands, providerDevOverrideRuntimeWarnings should be used.
+func (m *Meta) providerDevOverrideInitWarnings() tfdiags.Diagnostics {
+	if len(m.ProviderDevOverrides) == 0 {
+		return nil
+	}
+	var detailMsg strings.Builder
+	detailMsg.WriteString("The following provider development overrides are set in the CLI configuration:\n")
+	for addr, path := range m.ProviderDevOverrides {
+		detailMsg.WriteString(fmt.Sprintf(" - %s in %s\n", addr.ForDisplay(), path))
+	}
+	detailMsg.WriteString("\nThis command is not necessary for using provider development overrides in other commands and may behave unexpectedly. Any overrides for non-existant provider sources will cause an error in this command.")
```

Comment:
I think this message would benefit from more direct phrasing, though I don't have a single suggestion, just some messy ideas: 

- "terraform init will always error when ...", 
- "Do not run init when provider dev... "

Reply:
Happy to adjust! Maybe something along the lines of the below?

Skip this command when using provider development overrides. It is not necessary and may error unexpectedly.

Reply:
Very nice, that's much better than anything I came up with! 

I would prefer if it says "init" instead of "this command" (or if we refer to "init" a sentence or so before "this command"). The way it currently reads to me is that "this command" is referring to the `Provider development overrides`, not `init`:

```
Warning: Provider development overrides are in effect

The following provider development overrides are set in the CLI configuration:
 - mildwonkey/random in /Users/kristin/go/src/github.com/mildwonkey/terraform-provider-random

This command is not necessary ...
```

Reply:
My apologies, I was getting hung up on not including `init` because that is the style needed for the other provider developer warning functions as they apply to multiple commands. Certainly safe in this case to refer directly to `init` since it is only used there! 👍 

```suggestion
	detailMsg.WriteString("\nSkip terraform init when using provider development overrides. It is not necessary and may error unexpectedly.")
```

Reply:
Love it!

Reply:
(And no need to apologize!)